engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/internal/PairwiseDoubleRingBuffer.java

```diff
@@ -0,0 +1,297 @@
+    private int capacity;
+    private int chunkSize;
+
+    private int head;
+    private int tail;
+
+    @FunctionalInterface
+    public interface DoubleFunction {
+        /**
+         * Applies this function to the given arguments.
+         *
+         * @param a the first function argument
+         * @param b the second function argument
+         * @return the function result
+         */
+        double apply(double a, double b);
+    }
+
+    /**
+     * Create a ring buffer for Double values that will perform pairwise evaluation of the internal data values using an efficient binary-tree implementation to compute only changed values.  The buffer will grow exponentially as items are pushed into it but will not shrink as values are removed
+     *
+     * @param initialSize the minimum size for the structure to hold
+     * @param emptyVal an innocuous value that will not affect the user-provided function results. for example, 0.0f for performing addition/subtraction, 1.0f for performing multiplication/division
+     * @param pairwiseFunction the user provided function for evaluation, takes two double parameters and returns a double. This function will be applied repeatedly to pairs of data values until the final result is available
+     */
+    public PairwiseDoubleRingBuffer(int initialSize, double emptyVal, DoubleFunction pairwiseFunction) {
+        // increase to next power of two
+        this.capacity = Integer.highestOneBit(initialSize) * 2;
+        this.chunkSize = capacity * 2;
+        this.storageChunk = WritableDoubleChunk.makeWritableChunk(chunkSize);
+        this.dirtyIndices = new TIntArrayList(chunkSize);
+        this.pairwiseFunction = pairwiseFunction;
+        this.emptyVal = emptyVal;
+
+        this.storageChunk.fillWithValue(0, chunkSize, emptyVal);
+        this.head = this.tail = this.capacity;
+        this.allDirty = false;
+    }
+
+    private void evaluateRangeFast(int start, int end) {
+        // everything in this range needs recomputed
+        for (int left = start & 0xFFFFFFFE; left < end; left += 2) {
+            final int right = left + 1;
+            final int parent = left / 2;
+
+            // load the data values
+            final double leftVal = storageChunk.get(left);
+            final double rightVal = storageChunk.get(right);
+
+            // compute & stpre
+            final double computeVal = pairwiseFunction.apply(leftVal, rightVal);
+            storageChunk.set(parent, computeVal);
+
+            // mark the parent dirty
+            dirtyIndices.add(parent);
+        }
+    }
+
+    public double evaluate() {
+        // if all dirty, recompute all values
+        if (allDirty) {
+            if (head < tail) {
+                evaluateRangeFast(head, tail);
+            } else {
+                evaluateRangeFast(head, chunkSize);
+                evaluateRangeFast(capacity, tail);
+            }
+        } else {
+            // sort so consecutive values are adjacent
```

Comment:
Each level of the tree is independent but the initial set of `dirtyIndices` might be partially unsorted because of the wrap-around in the ring-buffer.  After additional thought, processing unsorted may be better while still being correct.

Reply:
Removed `sort()` call, verified tests still pass.